internal/postgres/toddl.go

```diff
@@ -0,0 +1,196 @@
+		}
+	}
+	switch id {
+	case "bool":
+		maxExpectedMods(0)
+		return ddl.Bool{}, nil
+	case "bigserial":
+		maxExpectedMods(0)
+		return ddl.Int64{}, []schemaIssue{serial}
+	case "bpchar": // Note: Postgres internal name for char is bpchar (aka blank padded char).
+		maxExpectedMods(1)
+		if len(mods) > 0 {
+			return ddl.String{ddl.Int64Length{mods[0]}}, nil
+		}
+		// Note: bpchar without length specifier is equivalent to bpchar(1)
+		return ddl.String{ddl.Int64Length{1}}, nil
+	case "bytea":
+		maxExpectedMods(0)
+		return ddl.Bytes{ddl.MaxLength{}}, nil
+	case "date":
+		maxExpectedMods(0)
+		return ddl.Date{}, nil
+	case "float8":
+		maxExpectedMods(0)
+		return ddl.Float64{}, nil
+	case "float4":
+		maxExpectedMods(0)
+		return ddl.Float64{}, []schemaIssue{widened}
+	case "int8":
+		maxExpectedMods(0)
+		return ddl.Int64{}, nil
+	case "int4":
+		maxExpectedMods(0)
+		return ddl.Int64{}, []schemaIssue{widened}
+	case "int2":
+		maxExpectedMods(0)
+		return ddl.Int64{}, []schemaIssue{widened}
+	case "numeric": // Map all numeric types to float64.
+		maxExpectedMods(2)
+		if len(mods) > 0 && mods[0] <= 15 {
+			// float64 can represent this numeric type faithfully.
+			// Note: int64 has 53 bits for mantissa, which is ~15.96
+			// decimal digits.
+			return ddl.Float64{}, []schemaIssue{numericThatFits}
+		}
+		return ddl.Float64{}, []schemaIssue{numeric}
+	case "serial":
+		maxExpectedMods(0)
+		return ddl.Int64{}, []schemaIssue{serial}
+	case "text":
+		maxExpectedMods(0)
+		return ddl.String{ddl.MaxLength{}}, nil
+	case "timestamptz":
+		maxExpectedMods(1)
+		return ddl.Timestamp{}, nil
+	case "timestamp":
+		maxExpectedMods(1)
+		// Map timestamp without timezone to Spanner timestamp.
+		return ddl.Timestamp{}, []schemaIssue{timestamp}
+	case "varchar":
+		maxExpectedMods(1)
+		if len(mods) > 0 {
+			return ddl.String{ddl.Int64Length{mods[0]}}, nil
+		}
+		return ddl.String{ddl.MaxLength{}}, nil
+	}
+	return ddl.String{ddl.MaxLength{}}, []schemaIssue{noGoodType}
+}
+
+func prType(ty schema.Type) string {
+	s := ty.Id
+	if len(ty.Mods) > 0 {
+		var l []string
+		for _, x := range ty.Mods {
+			l = append(l, strconv.FormatInt(x, 10))
+		}
+		s = fmt.Sprintf("%s(%s)", s, strings.Join(l, ","))
+	}
+	if len(ty.ArrayBounds) > 0 {
+		l := []string{s}
+		for _, x := range ty.ArrayBounds {
+			if x == -1 {
+				l = append(l, "[]")
+			} else {
+				l = append(l, fmt.Sprintf("[%d]", x))
+			}
+		}
+		s = strings.Join(l, "")
+	}
+	return s
+}
+
+func quoteIfNeeded(s string) string {
+	for _, r := range s {
+		if unicode.IsLetter(r) || unicode.IsDigit(r) || unicode.IsPunct(r) {
+			continue
+		}
+		return strconv.Quote(s)
+	}
+	return s
+}
+
+func cvtPrimaryKeys(conv *Conv, srcTable string, srcKeys []schema.Key) []ddl.IndexKey {
+	var spKeys []ddl.IndexKey
+	for _, k := range srcKeys {
+		spCol, err := GetSpannerCol(conv, srcTable, k.Column, true)
+		if err != nil {
+			conv.unexpected(fmt.Sprintf("Can't map key for table %s", srcTable))
```

Comment:
If there is an error, what will `spCol` be? An empty string ""? We probably should not use this to create an index key.

Reply:
Good catch.
Added continue.